BHoM_Adapter/AdapterActions/Push.cs

```diff
@@ -35,33 +35,78 @@ namespace BH.Adapter
         /******************************************************/
         /* These methods represent Actions that the Adapter can complete. 
            They are publicly available in the UI as individual components, e.g. in Grasshopper, under BHoM/Adapters tab. */
 
-        [Description("Pushes input objects using either the 'Full CRUD', 'CreateOnly' or 'UpdateOnly', depending on the PushType.")]
-        public virtual List<object> Push(IEnumerable<object> objects, string tag = "", PushType pushType = PushType.AdapterDefault, ActionConfig actionConfig = null)
+        [Description("Performs a set up for the ActionConfig of the Push Action.")]
+        public virtual bool SetupPushConfig(ActionConfig actionConfig, out ActionConfig pushConfig)
         {
-            bool success = true;
+            // If null, set the actionConfig to a new ActionConfig.
+            pushConfig = actionConfig == null ? new ActionConfig() : actionConfig;
 
-            // ----------------------------------------//
-            //                 SET-UP                  //
-            // ----------------------------------------//
+            return true;
+        }
+
+        [Description("Performs a set up for the objects to be sent to the Push Action.")]
+        public virtual bool SetupPushObjects(IEnumerable<object> objects, ActionConfig actionConfig, out IEnumerable<object> objectsToPush)
+        {
+            // Process the objects (verify they are valid; DeepClone them, wrap them, etc).
+
+            // Read ActionConfig.
+            // If ActionConfig has a value for `WrapNonBHoMObjects`, it has precedence over the default value in AdapterSettings.
+            bool wrapNonBHoMObjects = m_AdapterSettings.WrapNonBHoMObjects;
+            if (actionConfig != null)
+                wrapNonBHoMObjects = actionConfig.WrapNonBHoMObjects;
+
+            // Object verification. 
+            objectsToPush = new List<IBHoMObject>();
+
+            // Verify that the input objects are IBHoMObjects.
+            if (objects.OfType<IBHoMObject>().Count() != objects.Count() & !wrapNonBHoMObjects)
+            {
+                Engine.Reflection.Compute.RecordWarning("Only non-null BHoMObjects are supported by the default Push. " + // = you can override if needed; 
+                    "\nConsider specifying actionConfig['WrapNonBHoMObjects'] to true.");
+            }
 
+            // Wrap non-BHoM objects into a Custom BHoMObject to make them compatible with the CRUD.
+            if (wrapNonBHoMObjects)
+                objectsToPush = WrapNonBHoMObjects(objects);
+            else
+                objectsToPush = objects.OfType<IBHoMObject>();
+
+            // Clone the objects for immutability in the UI. CloneBeforePush should always be true, except for very specific cases.
+            if (m_AdapterSettings.CloneBeforePush)
+                objectsToPush = objectsToPush.Select(x => x.DeepClone());
+
+            return true;
+        }
```

Comment:
As for the adapter verification of the Move, not sure this is a setup. This is _not_ simply converting an object into a request or similar, this is doing stuff I would say is part of the actual push, controlled via action configs and similar. Would not put all this in this separate set up method.

Reply:
Not that I disagree, but this method was already written as a separate method, [`ProcessObjectsForPush()`](https://github.com/BHoM/BHoM_Adapter/blob/master/BHoM_Adapter/AdapterActions/_PushMethods/ProcessObjectsForPush.cs).

It seemed a good opportunity to formalise it better, but then again I understand what you're saying here: the logic that this method contains is strictly related to the Push logic, and therefore it makes more sense as more "embedded" in the Push than separated.

I will restore it as it was before.